<?xml version="1.0"?>

  <robot name="my_rigid_arm" xmlns:xacro="http://ros.org/wiki/xacro">

  <material name="black">
    <color rgba="0 0 0 1"/>
  </material>

  <material name="purple">
    <color rgba="1 0 1 1"/>
  </material>


  <link name="base_link" />

  <xacro:macro name="addSegment" params="parent child color leng">

  <!-- create the next link -->
  <link name="${child}">
    <visual>
      <geometry>
            <cylinder radius="0.025" length="${leng}"/>
      </geometry>
      <origin rpy="0 0 0" xyz="0 0 ${leng/2}"/>
      <material name="${color}"/>
    </visual>
  </link>

  <!-- setup the coordinate frame -->
  <joint name="${parent}${child}" type="floating" >
      <origin xyz="0 0 0" />
      <parent link="${parent}" />
      <child link="${child}" />
  </joint> 
  </xacro:macro>

  <xacro:macro name="addMarker" params="name">

    <link name="${name}">
      <visual>
        <geometry>
          <sphere radius="0.04"/>
        </geometry>
        <material name="green">
              <color rgba="0 1 0 1"/>
        </material>
      </visual>
    </link>

    <!-- setup the coordinate frame -->
  <joint name="base_link${name}" type="floating" >
      <origin xyz="0 0 0" />
      <parent link="base_link" />
      <child link="${name}" />
  </joint> 
  </xacro:macro>

  <xacro:addSegment parent="base_link" child="11" color="black" leng="0.01555"/> 
<xacro:addSegment parent="base_link" child="12" color="black" leng="0.01555"/> 
<xacro:addSegment parent="base_link" child="13" color="black" leng="0.01555"/> 
<xacro:addSegment parent="base_link" child="14" color="black" leng="0.01555"/> 
<xacro:addSegment parent="base_link" child="21" color="purple" leng="0.02"/> 
<xacro:addSegment parent="base_link" child="22" color="purple" leng="0.02"/> 
<xacro:addSegment parent="base_link" child="23" color="purple" leng="0.02"/> 
<xacro:addSegment parent="base_link" child="24" color="purple" leng="0.02"/> 
<xacro:addSegment parent="base_link" child="31" color="black" leng="0.04"/> 
<xacro:addSegment parent="base_link" child="32" color="black" leng="0.04"/> 
<xacro:addSegment parent="base_link" child="41" color="purple" leng="0.08"/> 
<xacro:addSegment parent="base_link" child="51" color="black" leng="0.016"/> 
<xacro:addSegment parent="base_link" child="52" color="black" leng="0.016"/> 
<xacro:addSegment parent="base_link" child="53" color="black" leng="0.016"/> 
<xacro:addSegment parent="base_link" child="54" color="black" leng="0.016"/> 
<xacro:addSegment parent="base_link" child="55" color="black" leng="0.016"/> 
<xacro:addSegment parent="base_link" child="61" color="purple" leng="0.04"/> 
<xacro:addSegment parent="base_link" child="62" color="purple" leng="0.04"/> 
<xacro:addSegment parent="base_link" child="71" color="black" leng="0.02"/> 
<xacro:addSegment parent="base_link" child="72" color="black" leng="0.02"/> 
<xacro:addSegment parent="base_link" child="73" color="black" leng="0.02"/> 
<xacro:addSegment parent="base_link" child="74" color="black" leng="0.02"/> 
<xacro:addSegment parent="base_link" child="81" color="purple" leng="0.04"/> 
<xacro:addSegment parent="base_link" child="82" color="purple" leng="0.04"/> 

<xacro:addMarker name='marker_0'/> 
<xacro:addMarker name='marker_1'/> 
</robot> 
